src/csharp/Microsoft.Spark.Worker/Command/SqlCommandExecutor.cs

```diff
@@ -294,13 +296,87 @@ public object Run(int splitId, object input)
         }
     }
 
+    internal class ArrowBasedCommandExecutor : SqlCommandExecutor
+    {
+        protected internal override CommandExecutorStat ExecuteCore(Stream inputStream, Stream outputStream, SqlCommand[] commands)
+        {
+            throw new NotImplementedException();
+        }
+        protected IEnumerable<RecordBatch> GetInputIterator(Stream inputStream)
+        {
+            using (var reader = new ArrowStreamReader(inputStream, leaveOpen: true))
+            {
+                RecordBatch batch;
+                bool returnedResult = false;
+                while ((batch = reader.ReadNextRecordBatch()) != null)
+                {
+                    yield return batch;
+                    returnedResult = true;
+                }
+
+                if (!returnedResult)
+                {
+                    // When no input batches were received, return an empty RecordBatch
+                    // in order to create and write back the result schema.
+
+                    int columnCount = reader.Schema.Fields.Count;
+                    var arrays = new IArrowArray[columnCount];
+                    for (int i = 0; i < columnCount; ++i)
+                    {
+                        IArrowType type = reader.Schema.GetFieldByIndex(i).DataType;
+                        arrays[i] = ArrowArrayHelpers.CreateEmptyArray(type);
+                    }
+
+                    yield return new RecordBatch(reader.Schema, arrays, 0);
+                }
+            }
+        }
+    }
+
+    internal class ArrowOrDataFrameSqlCommandExecutor : ArrowBasedCommandExecutor
+    {
+        protected internal override CommandExecutorStat ExecuteCore(
+            Stream inputStream,
+            Stream outputStream,
+            SqlCommand[] commands)
+        {
+            bool useDataFrameCommandExecutor = false;
+            bool useArrowSqlCommandExecutor = false;
+            foreach (SqlCommand command in commands)
+            {
+                WorkerFunction workerFunc = command.WorkerFunction;
+                if (workerFunc is DataFrameWorkerFunction dataFrameWorkedFunc)
+                {
+                    useDataFrameCommandExecutor = true;
+                }
+                else
+                {
+                    useArrowSqlCommandExecutor = true;
+                }
+            }
+            if (useDataFrameCommandExecutor && useArrowSqlCommandExecutor)
+            {
+                // Mixed mode. Not supported
+                throw new NotSupportedException("Combined Arrow and DataFrame style commands are not supported");
+            }
+            SqlCommandExecutor executor = null;
+            if (useDataFrameCommandExecutor)
+            {
+                executor = new DataFrameSqlCommandExecutor();
```

Comment:
Do we really need to create a new object here? Can we accomplish the same thing with just calling different methods on the current class?

Reply:
Testing this.

Reply:
Yup, good idea. I was going to move all the common methods to the base class, but this is a step further and better.